### ISSUE TITLE
Mean estimate larger than maximum value

### ISSUE THREAD

**Issue by seaaan** [stmcg/estmeansd#1]
Hello! Thank you for your work in developing these methods and implementing them in this package. The package is easy to use and super helpful! 

I have been testing your methods for a meta-analysis I am going to perform. I am testing using the data sets where I have the complete data so I can compare the estimates provided by your algorithms to the true values, using data of similar structure to what I'll have in the meta-analysis. I noticed that in some cases I am getting mean estimates that are larger than the maximum value supplied to the function and I wonder if I am doing something wrong. 

For example: 
```
qe.mean.sd(min.val = 1, q1.val = 1, med.val = 3, q3.val = 9, max.val = 300, n = 5)
$est.mean
[1] 5189016798

$est.sd
[1] 2.69259e+19
```

In this case I have only five values in my data, most of which are very low and one of which is ~100-fold higher. The estimate of the mean is astronomical and much higher than the maximum value, which should be impossible. The mean drops below the maximum value only at `n = 22`. Is there some way I can adjust the fitting algorithm to better handle cases like this? Or am I using the function wrong? 

Another example, this time with the other function: 
```
set.seed(17)
bc.mean.sd(min.val = 0.3, q1.val = 1, med.val = 20, q3.val = 150, max.val = 2500, n = 30)
$est.mean
[1] 4204.749

$est.sd
[1] 84380.73
```
In this case the function usually gives a mean estimate within the bounds of the inputs, but one some runs (such as with 17 as the seed), the mean estimate exceeds the maximum value.

Again, thanks for this great package and set of algorithms! I'd be grateful for any advice you can offer.

**Comment by stmcg** [comment]
Hi,

First off, thanks for posting and apologies for getting to this so late. I just realized that I didn't have this repository on watch and so I missed the notification.

It seems that you are using the functions correctly and, taking a quick look into these specific cases, I don't believe there is a bug in the software causing these unreasonable estimates either. In the applications we had in mind when developing the methods, we were considering somewhat large samples sizes (in our simulation study, the lowest n was 25 for instance) and we weren't considering the case of extremely large maximum values (eg, 100-fold higher as you mention). While the sample mean should of course be less than or equal to the maximum value, the estimation algorithms don't enforce this and are susceptible to this when sample sizes are small and the maximum value is extremely large.

Specifically, for the QE method, it looks like the optimization algorithm (ie, when minimizing the distance between the sample and distribution quantiles) is indeed converging for most of the candidate distributions in your example (small note: I see that the default parameter bounds for the sigma parameter in the normal distribution aren't wide enough for this example). Since we currently give equal weight to each of the 5 quantiles in the optimization algorithm, the very large maximum value is strongly affecting the parameter estimates. Some initial thoughts on solutions:

1.  Down-weight the maximum value in the optimization algorithm. The question then becomes finding a principled way to weight the quantiles, which we did consider early on but didn't find a suitable general solution. 
2. I suspect you'll get better results when using the L1 norm (instead of the L2 norm as currently used) in the optimization algorithm in the case of a very large maximum value. 
3. Simpler than 1 and 2, another idea would just be to only use the median and first and third quartiles when estimating the sample mean and standard deviation. I suppose this is just a special case of 1 where a weight of zero is given to the maximum and minimum.

I'm not as familiar with the BC approach and so I don't have specific suggestions for this approach in this case. 

More generally, it may be worth considering meta-analyzing the study-specific medians or difference of medians instead of the means / difference of means given that you have extremely skewed data. We wrote a few papers ([here](https://onlinelibrary.wiley.com/doi/10.1002/bimj.201900036) and [here](https://onlinelibrary.wiley.com/doi/abs/10.1002/sim.8013)) with corresponding R package `metamedian` proposing some approaches for meta-analyzing medians when studies report medians along with first and third quartiles and/or minimum and maximum values.

Best,
Sean

**Comment by seaaan** [comment]
Thank you for getting back to me and for your helpful suggestions! I was
not aware of the metamedian package and your papers on that topic. I will
look into those as they look helpful.

With regard to your point 2, could you explain how to change from the L1 to
L2 norm as well as the meaning/effect of that change?

Thanks again!
Sean

**Comment by stmcg** [comment]
Glad to hear that the median papers look helpful for your analyses.

As for point 2:

I'd describe the QE method as follows. We start by assuming some parametric family of distributions. Let's say the normal distribution to fix ideas. Then we estimate the parameters of the underlying normal distribution (i.e., the mean and standard deviation) by minimizing the "distance" between the sample quantiles reported in the study and the corresponding distribution quantiles (i.e., the 1/n, 25, 50, 75, and 1-1/n quantiles). We do this for several different parametric families of distributions and select the one with the best fit to estimate the sample mean and standard deviation.

Now, the "distance" can be defined in several ways. In our paper and software, we used what I was referring to as the L2 norm. This is the sum of the squared differences between the sample and distribution quantiles. The L1 norm would instead be the sum of the absolute value of the difference between the sample and distribution quantiles. 

Suppose I have the sample quantiles from your first example. Also, suppose that in the current step in the optimization algorithm, the corresponding distribution quantiles are: 0.8, 1.3, 3.7, 11, 290. 

The L2 norm would be (0.8 - 1)^2 + (1.3 - 1)^2 + (3.7 - 3)^2 + (11 - 9)^2 + (290-300)^2 = 104.62. The term for the maximum value (i.e., (290-300)^2) contributes a distance of 100 to the objective function, which dominates all the other terms which only add up to 4.62.

The L1 norm would be |0.8 - 1| + |1.3 - 1| + |3.7 - 3| + |11 - 9| + |290-300| = 13.2. The term for the maximum value (i.e., |290-300|) contributes a distance of 10 to the objective function and the other terms contribute a distance of 3.2. So, the contribution from the maximum value doesn't dominate the other terms as much in the L1 case compared to the L2 case.

Using the L1 norm would involve a small change the objective functions in `qe.fit`. For instance, for the normal distribution, one would have to change the current objective function

```
S.theta.norm <- function(theta) {
        summand <- sum((stats::qnorm(p = probs, mean = theta[1], 
            sd = theta[2]) - quants)^2)
}
```
to 
```
S.theta.norm <- function(theta) {
        summand <- sum(abs(stats::qnorm(p = probs, mean = theta[1], 
            sd = theta[2]) - quants))
}
```
and similarly for the objective functions for the other distributions. 

I applied the method using the L1 norm to your first example out of curiosity but unfortunately still got crazy values (estimated mean = 9903849, estimated sd = 3.293221e+13). 

I'm a bit hesitant on adding a feature to the package for using the L1 norm because we haven't explored how this estimator behaves. But I'd be happy to elaborate more if anything is unclear.
